feat: savings plugin and lido integration

NoblePaladin opened the pull request:
## Description

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Closes #5022

Please note, before merging. Please approve and review the workflow with Goerli.
Once it is approved. I will update it to use the Mainnet Ethereum Chain.

The change to Ethereum Mainnet just requires one line of code changed (`targetChainId` points to the protocols `chainId`, as seen in [SavingsForm.tsx](https://github.com/NoblePaladin/Maskbook/blob/%40savings/lido/packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx#L27))

## Type of change

<!-- Please remove options that are not relevant. -->

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Previews

https://user-images.githubusercontent.com/94978666/147374843-0ab38dc3-931a-47ae-883b-a7acaa88a7f8.mp4

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [x] I have configured those APIs with CORS headers to let extension requests get passed. <!-- If you don't have permission to modify the server, please let us know it. -->
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [x] I have delegated all web requests to the background service via the internal RPC bridge.

NoblePaladin commented:
![Screenshot from 2021-12-24 19-32-52](https://user-images.githubusercontent.com/94978666/147374926-a2c7a879-eab6-4dc4-9284-d3528067372c.png)

I also want to note that the "Withdraw" functionality, points to the Swap Plugin. Because ETH 2.0 Beacon Chain for Lido hasn't been implemented. Therefore, there is no actual way to "unstake" stETH.

Jack-Works commented on packages/mask/src/plugins/Savings/types.ts:
```diff
@@ -0,0 +1 @@
+export {}
```
If you don't need this file, you can remove it

Jack-Works commented on packages/mask/src/plugins/Savings/context.ts:
```diff
@@ -0,0 +1,3 @@
+import { createContext } from 'react'
+
+export const SnapshotContext = createContext<never>(null!)
```
Why you need a context that provides nothing?

septs reviewed (changes requested):
plz rename the pr title.

NoblePaladin commented:
@septs, made sure to update the pull request to use the correct pull request title.
@Jack-Works, I've removed both files since neither were needed at the moment.

I've also added some smaller updates such updating the cspell `ignoreWords` list as well too.
I am going to check on this in a bit to see if the CI workflows are running all correctly. And will revise the PR again if there are any issues running them.

NoblePaladin commented:
@septs, @Jack-Works,

I've spent some time improving the codebase's code structure. Making sure that it's more DRY.
However, since this is my first PR. I'd imagine there will be some feedback on implementation.

Can review the updated code when you get a chance.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -0,0 +1,102 @@
+
+    const currentChainId = useChainId()
+    const [chainId, setChainId] = useState<ChainId>(currentChainId)
+    const [tab, setTab] = useState('deposit' as 'deposit' | 'withdraw')
```
```suggestion
    const [tab, setTab] = useState<'deposit' | 'withdraw'>('deposit')
```

guanbinrui replied:
We can have an enum for this.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -0,0 +1,102 @@
+    const currentChainId = useChainId()
+    const [chainId, setChainId] = useState<ChainId>(currentChainId)
+    const [tab, setTab] = useState('deposit' as 'deposit' | 'withdraw')
+    const [selectedProtocol, setSelectedProtocol] = useState(-1 as number)
```
```suggestion
    const [selectedProtocol, setSelectedProtocol] = useState(-1)
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -0,0 +1,102 @@
+    const account = useAccount()
+
+    useEffect(() => {
+        ;(async () => {
```
Try [useAsync](https://github.com/streamich/react-use/blob/master/docs/useAsync.md).

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,157 @@
+    const { t } = useI18N()
+    const { classes } = useStyles()
+    const protocol = SavingsProtocols[selectedProtocol]
+    const targetChainId = 5 // change to `chainId` later
```
```suggestion
    const targetChainId = ChainId.Gorli // change to `chainId` later
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,157 @@
+        if (tokenAmount.isZero()) return t('plugin_trader_error_amount_absence')
+        if (isLessThan(inputAmount, 0)) return t('plugin_trade_error_input_amount_less_minimum_amount')
+        if (
+            new BigNumber(
```
Try `isLessThan` in `packages/web3-shared/base/utils/number.ts`.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/protocols/index.ts:
```diff
@@ -0,0 +1,20 @@
+        deposit: lidoDeposit,
+        withdraw: lidoWithdraw,
+        availableNetworks: [
+            { chainId: 1, chainName: 'Ethereum' },
+            { chainId: 5, chainName: 'Gorli' },
```
```suggestion
            { chainId: ChainId.Mainnet, chainName: 'Ethereum' },
            { chainId: ChainId.Gorli, chainName: 'Gorli' },
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/protocols/index.ts:
```diff
@@ -0,0 +1,20 @@
+        image: 'lido',
+        base: 'ETH',
+        pair: 'stETH',
+        apr: '5.3',
```
Why is the apr a fixed value? Should it fetch from contract or external APIs?

NoblePaladin replied:
The APR fetched is from `getLidoAPR()` in `SavingsDialog.tsx`.
However, there is still a CORS issue with the CORS anywhere plugin you suggested.

I'm pretty sure that Lido blocks any request to it's website that isn't it's origin.
Do you have any suggestions of where to fetch the actual APR value?

I can't seem to find a reliable source for this that is non trivial.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/protocols/LDOProtocol.tsx:
```diff
@@ -0,0 +1,102 @@
+    stEthContract: string
+}
+
+export const LidoContracts: { [key: number]: LidoContract } = {
```
Please move contract constants to `web3-constants`.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/protocols/LDOProtocol.tsx:
```diff
@@ -0,0 +1,102 @@
+
+export const LidoAPR = `https://cors.r2d2.to/?uri=https://stake.lido.fi/api/steth-apr`
+
+export function lidoContract(chainId: number, web3: Web3): Contract {
```
Here I recommend constraint saving protocols with `interface`. Each protocol should new a class to implement that interface. Just copy from the lido:

```
interface SavingProtocol {
    getAPR(): Promise<string>
    getBalance(): Promise<string>
    // ... more
}

class LidoProtocol implements SavingProtocol {
    // implementation
}
```

NoblePaladin commented:
@guanbinrui, check out the source code now.
Thanks so much for the feedback. Using class based implementations is so much cleaner.

Let me know when we got the green light.

guanbinrui commented on packages/dashboard/src/pages/Labs/index.tsx:
```diff
@@ -1,3 +1,4 @@
+
```
```suggestion
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -0,0 +1,84 @@
+
+    const currentChainId = useChainId()
+    const [chainId, setChainId] = useState<ChainId>(currentChainId)
+    const [tab, setTab] = useState<'deposit' | 'withdraw'>('deposit')
```
```ts
enum TabType {
    Deposit = 'deposit',
    Withdraw = 'withdraw',
}
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,158 @@
+    const web3 = useWeb3(false, targetChainId)
+    const account = useAccount()
+
+    const [inputAmount, setInputAmount] = useState('' as string)
```
```suggestion
    const [inputAmount, setInputAmount] = useState('')
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,158 @@
+        const tokenBalance =
+            (tab === 'deposit'
+                ? nativeTokenBalance
+                : (Number.parseFloat(protocol.balance) * Math.pow(10, 18)).toString()) || '0'
```
```suggestion
                : rightShift(protocol.balance, 18).integerValue().toFixed()
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,158 @@
+                ? nativeTokenBalance
+                : (Number.parseFloat(protocol.balance) * Math.pow(10, 18)).toString()) || '0'
+
+        if (new BigNumber(tokenBalance).isLessThan(tokenAmount)) {
```
```suggestion
        if (isLessThan(tokenBalance, tokenAmount)) {
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,158 @@
+    )
+
+    //#region form controls
+    const formattedBalance = Number.parseInt(nativeTokenBalance || '0', 16) / Math.pow(10, 18)
```
```suggestion
    const formattedBalance = leftShift(Number.parseInt(nativeTokenBalance || '0', 16), 18)
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,158 @@
+                    </Button>
+                    <div className={classes.inputCurrency}>
+                        <img src={IconURLS.eth} className={classes.inputImage} />
+                        <p>{tab === 'deposit' ? protocol.base : protocol.pair}</p>
```
```suggestion
                        <Typography variant="body2">{tab === 'deposit' ? protocol.base : protocol.pair}</Typography>
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,158 @@
+                <EthereumWalletConnectedBoundary
+                    ActionButtonProps={{ color: 'primary', classes: { root: classes.button } }}
+                    classes={{ connectWallet: classes.connectWallet, button: classes.button }}>
+                    <ActionButton
```
Refactor it to use `ActionButtonPromise`.

guanbinrui commented on packages/web3-constants/evm/rpc.json:
```diff
@@ -1,14 +1,14 @@
-        "Rinkeby": ["https://rinkeby.infura.io/v3/d74bd8586b9e44449cef131d39ceeefb"],
-        "Kovan": ["https://kovan.infura.io/v3/d74bd8586b9e44449cef131d39ceeefb"],
-        "Gorli": ["https://goerli.infura.io/v3/d74bd8586b9e44449cef131d39ceeefb"],
+        "Mainnet": ["https://mainnet.infura.io/v3/a98fb9412dcc41048e32586addfe3d29"],
```
Plz revert these changes.

guanbinrui commented on packages/mask/src/plugins/Savings/types.ts:
```diff
@@ -0,0 +1,25 @@
+    apr: string
+    balance: string
+
+    getContract(chainId: number, web3: Web3): Contract
+    getApr(): Promise<string>
+    getBalance(chainId: number, web3: Web3, account: string): Promise<string>
+    deposit(account: string, chainId: number, web3: Web3, value: BigNumber): Promise<boolean>
+    withdraw(account: string, chainId: number, web3: Web3, value: BigNumber): Promise<boolean>
```
```suggestion
    deposit(account: string, chainId: number, web3: Web3, value: BigNumber.Value): Promise<boolean>
    withdraw(account: string, chainId: number, web3: Web3, value: BigNumber.Value): Promise<boolean>
```

guanbinrui replied:
```suggestion
    getContract(chainId: ChainId, web3: Web3): Contract
    getApr(): Promise<string>
    getBalance(chainId: ChainId, web3: Web3, account: string): Promise<string>
    deposit(account: string, chainId: ChainId, web3: Web3, value: BigNumber.Value): Promise<boolean>
    withdraw(account: string, chainId: ChainId, web3: Web3, value: BigNumber.Value): Promise<boolean>
```

guanbinrui commented on packages/mask/src/plugins/Savings/types.ts:
```diff
@@ -0,0 +1,25 @@
+import type { Contract } from 'web3-eth-contract'
+
+export interface SavingsNetwork {
+    chainId: number
```
```suggestion
    chainId: ChainId
```

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/steth.abi.json:
```diff
@@ -0,0 +1,722 @@
+[
```
Move it to `web3-contracts`.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,158 @@
+
+    return (
+        <div className={classes.containerWrap}>
+            <h3 className={classes.title}>
```
Refactor to use `<Typography />`.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,158 @@
+                {tab === 'deposit' ? 'Stake' : 'Withdraw'} {protocol.name}
+            </h3>
+
+            <p className={classes.inputLabel}>
```
Refactor to use `<Typography />`.

NoblePaladin commented:
@guanbinrui, wait one second. Need to add the typography changes.

NoblePaladin commented:
@guanbinrui, just did a second review. I believe this should be good to go.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,165 @@
+    const { t } = useI18N()
+    const { classes } = useStyles()
+    const protocol = SavingsProtocols[selectedProtocol]
+    const targetChainId = chainId
```
`targetChainId` is unnecessary.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,165 @@
+import { ActionButtonPromise } from '../../../extension/options-page/DashboardComponents/ActionButton'
+
+export interface SavingsFormProps {
+    chainId: number
```
```suggestion
    chainId: ChainId
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,165 @@
+    }, [inputAmount, tokenAmount.toFixed(), formattedBalance])
+    //#endregion
+
+    const needsSwap = protocol.name === 'Lido' && tab === 'withdraw'
```
```suggestion
    const needsSwap = protocol.name === ProtocolType.Lido && tab === TabType.Withdraw
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,165 @@
+    )
+
+    //#region form controls
+    const formattedBalance = leftShift(Number.parseInt(nativeTokenBalance || '0', 16), 18)
```
Here it didn't format anything.

```suggestion
    const balance = leftShift(Number.parseInt(nativeTokenBalance || '0', 16), 18)
```

septs replied:
```suggestion
    const formattedBalance = leftShift(new BigNumber(nativeTokenBalance || '0', 16), 18)
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,165 @@
+            <Typography variant="body1" className={classes.inputLabel}>
+                Balance:{' '}
+                {tab === TabType.Deposit
+                    ? formattedBalance
```
Refactor to use `formatBalance`.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,165 @@
+            </Typography>
+
+            <div className={classes.inputWrap}>
+                <TextField
```
Refactor to use `TokenAmountPanel` component. It's able to avoid negative numbers and alphabet characters.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsTable.tsx:
```diff
@@ -0,0 +1,108 @@
+}))
+
+export interface SavingsTableProps {
+    chainId: number
```
```suggestion
    chainId: ChainId
```

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsTable.tsx:
```diff
@@ -0,0 +1,108 @@
+                                variant="contained"
+                                color="primary"
+                                onClick={() => setSelectedProtocol(protocol.id)}>
+                                {tab === 'deposit' ? 'Deposit' : 'Withdraw'}
```
```suggestion
                                {tab === TabType.Deposit ? 'Deposit' : 'Withdraw'}
```

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,120 @@
+
+    public async getApr() {
+        try {
+            const LidoAprUrl = `https://cors.r2d2.to/?uri=https://stake.lido.fi/api/steth-apr`
```
```suggestion
            const LidoAprUrl = `https://cors.r2d2.to/?https://stake.lido.fi/api/steth-apr`
```

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,120 @@
+        try {
+            const contract = this.getContract(chainId, web3)
+            const balance = await contract.methods.balanceOf(account).call()
+            const formattedBalance = (balance / Math.pow(10, 18)).toLocaleString(undefined, {
```
UI formatting shouldn't be concerned here. Otherwise, for each protocol, they need to implement a formatting logic. You could delay formatting it when rendering it into UI. Also, I recommend using the in repo `formatBalance` function to do this.

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,120 @@
+
+    public async deposit(account: string, chainId: number, web3: Web3, value: BigNumber) {
+        try {
+            const LidoReferralAddress = '0x278D7e418a28ff763eEeDf29238CD6dfcade3A3a'
```
Move to `web3-constants`.

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,120 @@
+            await contract.methods.submit(LidoReferralAddress).send({
+                from: account,
+                value,
+                gasLimit: 2100000,
```
It would help if you estimated before sending the transaction.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,165 @@
+        <div className={classes.containerWrap}>
+            <Typography variant="h3" className={classes.title}>
+                <img src={IconURLS.lido} className={classes.titleImage} />
+                {tab === TabType.Deposit ? 'Stake' : 'Withdraw'} {protocol.name}
```
Please read the [i18n-guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and trun dead code textures to i18n fields.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,165 @@
+                        complete={t('done')}
+                        disabled={validationMessage !== '' && !needsSwap}
+                        executor={async () => {
+                            if (tab === TabType.Deposit) {
```
Do you think any new `TabType` will be added in the future? If that's possible, you can refactor the code as:

```ts
switch (tab) {
    case TabType.Deposit:
        return ...
    case TabType.Withdraw:
        return ...
    default:
        unreachable(tab)
}
```

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,120 @@
+        { chainId: ChainId.Gorli, chainName: 'Gorli' },
+    ]
+
+    public getContract(chainId: number, web3: Web3): Contract {
```
```suggestion
    public getContract(chainId: ChainId, web3: Web3): Contract {
```

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    public async getBalance(chainId: number, web3: Web3, account: string) {
```
```suggestion
    public async getBalance(chainId: ChainId, web3: Web3, account: string) {
```

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,120 @@
+    ]
+
+    public getContract(chainId: number, web3: Web3): Contract {
+        const contract = new web3.eth.Contract(LidoStETHABI as AbiItem[], LidoContracts[chainId].stEthContract)
```
It's not type-safe to create a contract in this way. After making an ABI JSON in the `web3-contracts`, you can run `npm run start` in the CLI. It will create a TypeScript contract definition for the ABI. The compiled definition file contains a class named by the JSON filename. It can be leveraged by the `createContract`:

```ts
import LidoContract from '@masknet/web3-contracts/types/Lido'

const contract = createContract<LidoContract>(web3, address, ABI as AbiItem[])
```

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    public async deposit(account: string, chainId: number, web3: Web3, value: BigNumber) {
```
```suggestion
    public async deposit(account: string, chainId: ChainId, web3: Web3, value: BigNumber.Value) {
```

guanbinrui commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    public async withdraw(account: string, chainId: number, web3: Web3, value: BigNumber) {
```
```suggestion
    public async withdraw(account: string, chainId: ChainId, web3: Web3, value: BigNumber.Value) {
```

guanbinrui commented:
Hi @NoblePaladin, I also found that the UI implementation is different from the original design.

![image](https://user-images.githubusercontent.com/52657989/147804395-a1a56b78-d4ed-43a7-ac43-d1b970a8bf8a.png)

And here is the original one.

![image](https://user-images.githubusercontent.com/52657989/147804438-376e9b34-9784-4bc0-a83b-b2a6e6728ff7.png)

Learn more: https://www.figma.com/file/gVkQ67y285b4FXVV1KPThN/TwitterV1?node-id=17600%3A374185.

NoblePaladin commented:
@guanbinrui, there were some changes to the design for formality reasons. I did change some of the copy to reflect the original designs.

But as I mentioned prior. I believe "Deposit" is a much more accurate descriptor than "Supply".
Furthermore, I've also added a "Transaction Fee" section.

This was not in the original design.
You may want to do some changes to the design.

Other than design issues. This should be good to go.

guanbinrui commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,174 @@
+            tab === TabType.Deposit
+                ? await protocol.depositEstimate(account, targetChainId, web3, new BigNumber('1'))
+                : await protocol.withdrawEstimate(account, targetChainId, web3, new BigNumber('1'))
+        setEstimatedGas(gasEstimate)
```
Please make it returns the gasEstimate directly, `useState` is redundancy here.

NoblePaladin replied:
@guanbinrui,using the contract method. `estimateGas` uses:

```typescript
(method) PayableTransactionObject<string>.estimateGas(tx?: PayableTx | undefined): Promise<number>
```

Unfortunately there's no way to use it without async.

NoblePaladin commented:
The following are the latest screenshots. I made sure to accurately reflect the Figma designs.

![Screenshot from 2022-01-04 13-08-53](https://user-images.githubusercontent.com/94978666/148104757-eb1b392b-878c-400a-aee7-e4e1a17129e7.png)
![Screenshot from 2022-01-04 13-09-14](https://user-images.githubusercontent.com/94978666/148104758-4985b4e9-784e-4161-8f63-e25180d1586a.png)
![Screenshot from 2022-01-04 13-09-23](https://user-images.githubusercontent.com/94978666/148104762-3cae6336-15e2-4bca-ba1f-13c4330151ad.png)
![Screenshot from 2022-01-04 13-09-42](https://user-images.githubusercontent.com/94978666/148104763-c0544686-4e95-4cfb-b05b-2b2feef075df.png)
![Screenshot from 2022-01-04 13-09-51](https://user-images.githubusercontent.com/94978666/148104764-132347c1-1899-4cdb-a9a3-d8d68af22009.png)

septs commented:
plz resolve conflict.

septs commented on packages/mask/src/plugins/Savings/SNSAdaptor/IconURL.tsx:
```diff
@@ -0,0 +1,4 @@
+export const IconURLS: Readonly<Record<string, string>> = {
```
```suggestion
export const IconURLs: Readonly<Record<string, string>> = {
```

septs commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,199 @@
+                                            onSwapDialogOpen?.()
+                                            break
+                                        default:
+                                            if (
+                                                (await protocol.withdraw(account, targetChainId, web3, tokenAmount)) ===
+                                                false
+                                            ) {
```
```suggestion
                                            if (!(await protocol.withdraw(account, targetChainId, web3, tokenAmount))) {
```

NoblePaladin commented:
@septs, I've fixed the merge conflict. I've also added those changes. Thank you!

AinuX commented:
@NoblePaladin 

https://user-images.githubusercontent.com/2582974/148919766-2b68b319-bbdc-4dd3-8542-cd38b841522b.mp4

1. Please give a "Depositing" state and disable the "Deposit ETH" button while processing deposit tx 
2. Fail to deposit ETH

septs commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -35,6 +36,31 @@ export function SavingsDialog({ open, onClose, onSwapDialogOpen }: SavingsDialog
         return networks.map((network) => getChainIdFromNetworkType(network))
     }, [])
 
+    const CategorizedProtocols = Object.keys(ProtocolCategory).map((category) => {
+        return {
+            category,
+            protocols: SavingsProtocols.filter(
+                (protocol) => protocol.category.toLowerCase() === category.toLowerCase(),
+            ),
+        }
+    })
```
```suggestion
    const CategorizedProtocols = Object.keys(ProtocolCategory).map((category) => ({
            category,
            protocols: SavingsProtocols.filter(
                (protocol) => protocol.category.toLowerCase() === category.toLowerCase(),
            ),
        }))
```

septs commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -35,6 +36,31 @@ export function SavingsDialog({ open, onClose, onSwapDialogOpen }: SavingsDialog
+        }
+
+        for (const protocol of categorizedProtocol.protocols) {
+            for (const network of protocol.availableNetworks) {
+                if (network.chainId === chainId) {
+                    return true
+                }
+            }
```
```ts
return categorizedProtocol.protocols.some(({ availableNetworks }) => availableNetworks.some(network => network.chainId === chainId))
```

see https://mdn.io/array-some

septs commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -35,6 +36,31 @@ export function SavingsDialog({ open, onClose, onSwapDialogOpen }: SavingsDialog
+        }
+    })
+
+    const MappableProtocols = CategorizedProtocols.filter((categorizedProtocol) => {
+        if (categorizedProtocol.protocols.length === 0) {
+            return false
+        }
+
+        for (const protocol of categorizedProtocol.protocols) {
+            for (const network of protocol.availableNetworks) {
+                if (network.chainId === chainId) {
+                    return true
+                }
+            }
+        }
+
+        return false
+    })
```
```ts
.filter((categorizedProtocol) => categorizedProtocol.protocols.some(({ availableNetworks }) => availableNetworks.some(network => network.chainId === chainId)))
```

see https://mdn.io/array-some

NoblePaladin replied:
Nice! Like this a lot!

AinuX commented:
@NoblePaladin 
https://user-images.githubusercontent.com/2582974/149283261-00348c7a-8854-47df-bfb6-60a0c5311875.mp4
3. When I complete deposit operation, and then click the back arrow, will go to the application list page rather than the savings plugin entry page.
4. If the withdraw function will point to swap plugin directly, so don't need a inputbox in withdraw page.
5. When go to swap plugin, it's better to bring the swap parameter like from stETH to ETH.

NoblePaladin commented:
@AinuX, I've added those additional changes! Thanks for the feedback. Check it out now and see if there's anything that needs to be updated.

AinuX commented:
@NoblePaladin 

6. Decimal issue

![image](https://user-images.githubusercontent.com/2582974/149731221-63d83fe8-3873-458d-a758-2850b527205e.png)

if the balance<0.000001, it's better to display "Balance: < 0.000001", like as below,
![image](https://user-images.githubusercontent.com/2582974/149731575-baf680fa-ca54-4115-a09b-1273077052d1.png)

AinuX commented:
@NoblePaladin 
7. Optimize data display 

- value ≥ 1, displays 6 decimal places
- 0.000001 ≤ value <1, displays 6 decimal places
- 0 < value < 0.000001, display < 0.000001
- value == 0, display 0


![image](https://user-images.githubusercontent.com/2582974/149883014-741a3aae-4bf7-4821-83c3-f5a00f5f5cd3.png)

septs commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsTable.tsx:
```diff
@@ -129,11 +129,14 @@ export function SavingsTable({ chainId, tab, mappableProtocols, setSelectedProto
-                                        : Number(protocol.balance).toLocaleString(undefined, {
-                                              maximumFractionDigits: 6,
-                                          })}
+                                    {Number(protocol.balance) !== 0
+                                        ? Number(protocol.balance) < 0.000001
+                                            ? '< 0.000001'
+                                            : Number(protocol.balance).toLocaleString(undefined, {
+                                                  minimumFractionDigits: 6,
+                                                  maximumFractionDigits: 6,
+                                              })
+                                        : 0}
```
plz this expression extract as function

ChrisCates replied:
@septs, I've updated it to:

```
const formattedBalance = formatBalance(balance, 18, 6)
```

That's my bad!

NoblePaladin commented:
@septs, I've updated it [here].(https://github.com/NoblePaladin/Maskbook/blob/%40savings/lido/packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts#L106)

to this:

```
const formattedBalance = formatBalance(balance, 18, 6)
```

However, should I be storing this balance as `BigNumber` and not a string?
And then running `formatBalance` in the table?

NoblePaladin commented:
@septs, @AinuX, so I've gone a bit further.

I added `decimals` as a type in the Savings Plugin Interface.
And I made sure to store the `balance` as a `BigNumber` from now on.

@septs, I believe this is the best implementation. Please let me know if this needs changes.
@AinuX, this implementation should cover the missing use cases.

UncleBill commented on packages/mask/src/extension/options-page/DashboardComponents/ActionButton.tsx:
```diff
@@ -136,7 +138,9 @@ export function ActionButtonPromise(props: ActionButtonPromiseProps) {
 
     useUpdateEffect(() => {
-        setState((prev) => (prev === 'init' ? prev : 'init'))
+        if (!noUpdateEffect) {
```
Missed `noUpdateEffect` in dependency array.

UncleBill commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -0,0 +1,113 @@
+        categorizedProtocol.protocols.some(({ availableNetworks }) =>
+            availableNetworks.some((network) => network.chainId === chainId),
+        ),
+    )
```
1. We don't capitalize normal variables.
`CategorizedProtocols` => `categorizedProtocols`,
`MappableProtocols` => `mappableProtocols`

2. Better to memoize them with `useMemo`
3. `CategorizedProtocols` is a middle variable, put it inside memorization for `MappableProtocols`

UncleBill commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -0,0 +1,113 @@
+
+    return (
+        <InjectedDialog
+            open={open || false}
```
```suggestion
            open={open}
```

UncleBill commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -0,0 +1,113 @@
+    const [tab, setTab] = useState<TabType>(TabType.Deposit)
+    const [selectedProtocol, setSelectedProtocol] = useState<ProtocolType | null>(null)
+
+    const { value: chains } = useAsync(async () => {
```
```suggestion
    const { value: chains = EMPTY_LIST } = useAsync(async () => {
```

`import { EMPTY_LIST } from '@masknet/web3-shared-evm`


Use a stable array to reduce re-rendering.

UncleBill commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -0,0 +1,113 @@
+                                chainId={chainId}
+                                setChainId={setChainId}
+                                classes={classes}
+                                chains={chains ?? []}
```
```suggestion
                                chains={chains}
```
Handle nullish case outside jsx above.

UncleBill commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,212 @@
+    )
+    //#endregion
+
+    //#region trade state
+    const {
+        tradeState: [{ inputToken, outputToken }, dispatchTradeStore],
+        allTradeComputed,
+    } = AllProviderTradeContext.useContainer()
+    //#endregion
```
```suggestion
    //#region trade state
    const {
        tradeState: [_, dispatchTradeStore],
    } = AllProviderTradeContext.useContainer()
    //#endregion
```

NoblePaladin commented:
@UncleBill, thanks for the feedback. I've updated accordingly.
So, in order to fulfill the balance formatting (as per @AinuX).

I needed to update [this](https://github.com/NoblePaladin/Maskbook/blob/%40savings/lido/packages/shared/src/wallet/FormattedBalance.tsx#L11-L40).

This adds the `minimumBalance` prop to the `FormattedBalance` component.
Would be nice to add that into the Trader plugin too later.

@AinuX, should double check this logic just in case. Since the previous commit. I accidently only had `formatBalance` and not all test cases covered.

septs commented on packages/shared/src/wallet/FormattedBalance.tsx:
```diff
@@ -13,13 +13,22 @@ export interface FormattedBalanceProps extends withClasses<'balance' | 'symbol'>
-    const formatted = formatter(value ?? '0', decimals, significant)
+    const { value, decimals, significant, symbol, minimumBalance, formatter = (value) => value } = props
+    const formatted =
+        Number(formatter(value ?? '0', decimals, significant)) === 0
```
plz save `formatter(value ?? '0', decimals, significant)` result as variable

it's re-calculating

septs replied:
plz use https://github.com/DimensionDev/Maskbook/blob/develop/packages/web3-shared/base/utils/number.ts

e.g: `isZero(balance)` (n === 0) /  `isLessThan(balance, minimumBalance)` (a < b)

septs commented on packages/shared/src/wallet/FormattedBalance.tsx:
```diff
@@ -13,13 +13,22 @@ export interface FormattedBalanceProps extends withClasses<'balance' | 'symbol'>
     decimals?: number
     significant?: number
     symbol?: string
+    minimumBalance?: number
```
```suggestion
    minimumBalance?: BigNumber.Value
```

NoblePaladin commented:
@septs, I've made sure that's more readable and also converted `formatted` to mutable via `let`.

septs commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsTable.tsx:
```diff
@@ -0,0 +1,169 @@
+                                        value={protocol.balance}
+                                        decimals={protocol.decimals}
+                                        significant={6}
+                                        minimumBalance={new BigNumber(Math.pow(10, protocol.decimals - 6))}
```
```suggestion
                                        minimumBalance={leftShift(10, protocol.decimals - 6)}
```

septs replied:
https://github.com/DimensionDev/Maskbook/blob/0d7baa815155a02e577281e60eb0e305de83b644/packages/web3-shared/base/utils/number.ts#L52

septs commented on packages/shared/src/wallet/FormattedBalance.tsx:
```diff
@@ -13,13 +14,20 @@ export interface FormattedBalanceProps extends withClasses<'balance' | 'symbol'>
-    const formatted = formatter(value ?? '0', decimals, significant)
+    const { value, decimals, significant, symbol, minimumBalance, formatter = (value) => value } = props
+    let formatted = formatter(value ?? '0', decimals, significant)
+    if (minimumBalance && value && formatted !== '0') {
```
```suggestion
    if (minimumBalance && value && !isZero(formatted)) {
```

septs commented on packages/shared/src/wallet/FormattedBalance.tsx:
```diff
@@ -13,13 +14,20 @@ export interface FormattedBalanceProps extends withClasses<'balance' | 'symbol'>
-    const formatted = formatter(value ?? '0', decimals, significant)
+    const { value, decimals, significant, symbol, minimumBalance, formatter = (value) => value } = props
+    let formatted = formatter(value ?? '0', decimals, significant)
+    if (minimumBalance && value && !isZero(formatted)) {
+        if (isLessThan(value, minimumBalance)) {
+            formatted = '<' + formatter(minimumBalance, decimals, significant)
+        }
+    }
```
```suggestion
    if (value && minimumBalance && !isZero(formatted) && isLessThan(value, minimumBalance)) {
        formatted = '<' + formatter(minimumBalance, decimals, significant)
    }
```

NoblePaladin commented:
@septs, give me a sec. Need to make sure everything passes lint:ci. Will update in a few.

NoblePaladin commented:
@septs, okay should be good now.

septs commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,211 @@
+                        </Typography>
+                    ) : (
+                        <Typography variant="body2" textAlign="right" className={classes.tokenValueUSD}>
+                            &#x2248; <FormattedCurrency value={tokenValueUSD} sign="$" formatter={formatCurrency} />
```
```suggestion
                            &asymp; <FormattedCurrency value={tokenValueUSD} sign="$" formatter={formatCurrency} />
```

UncleBill commented on packages/web3-constants/evm/debank.json:
```diff
@@ -15,10 +15,6 @@
-        "Boba": "boba",
-        "Metis": "metis",
-        "Optimistic": "op"
+        "Aurora_Testnet": ""
```
Please revert this change. Our proxy server needs this config.

UncleBill commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsTable.tsx:
```diff
@@ -0,0 +1,167 @@
+                                        value={protocol.balance}
+                                        decimals={protocol.decimals}
+                                        significant={6}
+                                        minimumBalance={leftShift(10, protocol.decimals - 6)}
```
You meant to use `rightShift`, right?
```suggestion
                                        minimumBalance={rightShift(10, protocol.decimals - 6)}
```

UncleBill commented on packages/mask/src/plugins/Savings/SNSAdaptor/IconURL.tsx:
```diff
@@ -0,0 +1,4 @@
+export const IconURLs: Readonly<Record<string, string>> = {
+    lido: new URL('./assets/lido.png', import.meta.url).toString(),
+    eth: new URL('./assets/eth.png', import.meta.url).toString(),
```
`eth.png` is too lower resolution, use an existed image or a higher resolution one.

UncleBill commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsDialog.tsx:
```diff
@@ -0,0 +1,122 @@
+            )
+    }, [chainId])
+
+    console.log(chainId, mappableProtocols)
```
Don't forget to remove debugging `console.log`, please check the other files as well.

UncleBill commented on packages/mask/src/components/shared/ApplicationBoard.tsx:
```diff
@@ -319,49 +336,58 @@ export function ApplicationBoard({ secondEntries, secondEntryChainTabs }: MaskAp
+                ) : null}
+                {isSavingsDialogOpen ? (
+                    <SavingsDialog
+                        open={isSavingsDialogOpen}
```
```suggestion
                        open
```

Please clean up `open` prop on other `Dialog` in this file as well, that will make code cleaner.

UncleBill commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,185 @@
+            )
+            const gasEstimate = await contract?.methods
+                .submit(
+                    getSavingsConstants(chainId).LIDO_REFERRAL_ADDRESS || '0x0000000000000000000000000000000000000000',
```
```suggestion
                    getSavingsConstants(chainId).LIDO_REFERRAL_ADDRESS || ZERO_ADDRESS
```

We have a constant `ZERO_ADDRESS` which you can import from `@masknet/web3-shared-evm` package. Please replace other literal zero addresses with it as well.

NoblePaladin commented:
@UncleBill, the only pending issue at the moment is how to handle the dialog for swaps properly.
As per `stETH` and the current state of Lido. There is no way to `withdraw` and you can only swap it to ETH.

I'm not sure if there is another way to handle this. Unless we refactor dialog states to be managed by Redux.
I am totally in favour of dialog states being handled by redux though!

Regardless, I've submitted updates. Please let me know if there are any other issues.

UncleBill commented on packages/mask/src/components/shared/ApplicationBoard.tsx:
```diff
@@ -350,52 +367,63 @@ export function ApplicationBoard({ secondEntries, secondEntryChainTabs }: MaskAp
+                            ) : null,
+                    )}
+                </section>
+                {isClaimAllDialogOpen ? (
+                    <ClaimAllDialog open={isClaimAllDialogOpen} onClose={onClaimAllDialogClose} />
+                ) : null}
+
+                {isSecondLevelEntryDialogOpen ? (
+                    <EntrySecondLevelDialog
+                        title={secondLevelEntryDialogTitle}
+                        open={isSecondLevelEntryDialogOpen}
+                        entries={secondLevelEntries}
+                        chains={secondLevelEntryChains}
+                        closeDialog={onSecondLevelEntryDialogClose}
+                    />
+                ) : null}
+                {isFindTrumanDialogOpen ? (
+                    <FindTrumanDialog open={isFindTrumanDialogOpen} onClose={onFindTrumanDialogClose} />
+                ) : null}
+
+                {isSwapDialogOpen ? <TraderDialog open={isSwapDialogOpen} onClose={onSwapDialogClose} /> : null}
+                {isSavingsDialogOpen ? (
+                    <SavingsDialog
+                        open={isSavingsDialogOpen}
+                        onClose={onSavingsDialogClose}
+                        onSwapDialogOpen={onSwapDialogOpen}
```
Since these dialogs are in the true branch of the ternary expression, you can omit the `isXxxx`.

```suggestion
                {isClaimAllDialogOpen ? (
                    <ClaimAllDialog open onClose={onClaimAllDialogClose} />
                ) : null}

                {isSecondLevelEntryDialogOpen ? (
                    <EntrySecondLevelDialog
                        title={secondLevelEntryDialogTitle}
                        open
                        entries={secondLevelEntries}
                        chains={secondLevelEntryChains}
                        closeDialog={onSecondLevelEntryDialogClose}
                    />
                ) : null}
                {isFindTrumanDialogOpen ? (
                    <FindTrumanDialog open onClose={onFindTrumanDialogClose} />
                ) : null}

                {isSwapDialogOpen ? <TraderDialog onClose={onSwapDialogClose} /> : null}
                {isSavingsDialogOpen ? (
                    <SavingsDialog
                        open
                        onClose={onSavingsDialogClose}
                        onSwapDialogOpen={onSwapDialogOpen}
```

Or you can just write the dialogs without ternary expression
```suggestion

                    <ClaimAllDialog open={isClaimAllDialogOpen} onClose={onClaimAllDialogClose} />

                    <EntrySecondLevelDialog
                        title={secondLevelEntryDialogTitle}
                        open={isSecondLevelEntryDialogOpen}
                        entries={secondLevelEntries}
                        chains={secondLevelEntryChains}
                        closeDialog={onSecondLevelEntryDialogClose}
                    />


                    <FindTrumanDialog open={isFindTrumanDialogOpen} onClose={onFindTrumanDialogClose} />


                 <TraderDialog open={isSwapDialogOpen} onClose={onSwapDialogClose} /> 
                    <SavingsDialog
                        open={isSavingsDialogOpen}
                        onClose={onSavingsDialogClose}
                        onSwapDialogOpen={onSwapDialogOpen}
```

UncleBill commented on packages/mask/src/components/shared/ApplicationBoard.tsx:
```diff
@@ -350,52 +367,63 @@ export function ApplicationBoard({ secondEntries, secondEntryChainTabs }: MaskAp
-                        setChainId={setChainId}
-                        classes={classes}
-                        chains={secondEntryChainTabs}
+        <TargetChainIdContext.Provider>
+            <AllProviderTradeContext.Provider>
```
I see you move these context providers from `packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx` here, do some of your components rely on them too? If so, you can wrap them with them too rather than wrap a high order component, and leave the providers in `TraderDialog` to make the code compact:


```tsx
 <TargetChainIdContext.Provider>
            <AllProviderTradeContext.Provider>
			<YourComponentThatRelysTradeContext />
            </AllProviderTradeContext.Provider>
 </TargetChainIdContext.Provider>
````

UncleBill commented on packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx:
```diff
@@ -0,0 +1,211 @@
+            <div className={classes.infoRow}>
+                <Typography variant="body1" className={classes.infoRowLeft}>
+                    <img src={IconURLs[protocol.image]} className={classes.rowImage} />
+                    {protocol.pair} {t('plugin_savings_apy')}%
+                </Typography>
+                <Typography variant="body1" className={classes.infoRowRight}>
+                    {protocol.apr}%
```
APR is different from APY. Do you need to align them here?

NoblePaladin commented:
@UncleBill, agree that we should remove the `open` prop altogether.

However, should be a separate PR.
One thing I do want to note. Is that also need to update the `InjectedDialog` component to not use the `open` prop.

For the time being. I've updated the `Savings` plugin accordingly.
Happy to open up a separate PR related to that.

UncleBill commented on packages/mask/src/plugins/Savings/protocols/LDOProtocol.ts:
```diff
@@ -0,0 +1,180 @@
+        }
+    }
+
+    public async getApy() {
+        try {
+            const LidoAprUrl = 'https://cors.r2d2.to/?https://stake.lido.fi/api/steth-apr'
+            const response = await fetch(LidoAprUrl)
+            const apy = await response.text()
+            this.apy = apy
+            return apy
+        } catch (error) {
+            console.log('LDO `getApy()` error', error)
+            // Default APR is 5.30%
+            this.apy = '5.30'
+            return '5.30'
+        }
+    }
```
According to the url, it should be APR instead.

You can learn the difference from https://www.americanexpress.com/en-us/banking/online-savings/articles/apy-vs-apr/

wenluomask commented:
Sorry to bother you again. Could you please continue to fix some UI bugs?

1. Since transaction is done, it will change to the Done button. However, the done button should be green;
<img width="845" alt="Screen Shot 2022-02-11 at 4 13 29 PM" src="https://user-images.githubusercontent.com/85149605/153558617-c15b7c77-bb90-49d6-80f1-02fd5c22d4d5.png">

2. For the network other than ETH, I hope the space under tab should be same as ETH;
Others:
<img width="1023" alt="Screen Shot 2022-02-11 at 4 17 44 PM" src="https://user-images.githubusercontent.com/85149605/153558892-262a8ff4-99e9-4c9c-846f-0b4bda38edae.png">
ETH:
<img width="926" alt="Screen Shot 2022-02-11 at 4 23 54 PM" src="https://user-images.githubusercontent.com/85149605/153558899-40ac7826-d641-4605-aff6-bb06d3e9a916.png">
